mqtt/policy/src/validator.rs

```diff
@@ -1,19 +1,53 @@
-use crate::{errors::Result, PolicyDefinition};
+use thiserror::Error;
+
+use crate::{PolicyDefinition, Statement};
 
 /// Trait to extend `PolicyBuilder` validation for policy definition.
 pub trait PolicyValidator {
+    /// The type of the validation error.
+    type Error;
+
     /// This method is being called by `PolicyBuilder` for policy definition
     /// while `Policy` is being constructed.
     ///
     /// If a policy definitions fails the validation, the error is returned.
-    fn validate(&self, definition: &PolicyDefinition) -> Result<()>;
+    fn validate(&self, definition: &PolicyDefinition) -> Result<(), Self::Error>;
 }
 
+/// Provides basic validation that policy definition elements are not empty.
 #[derive(Debug)]
 pub struct DefaultValidator;
 
 impl PolicyValidator for DefaultValidator {
-    fn validate(&self, _definition: &PolicyDefinition) -> Result<()> {
+    type Error = ValidatorError;
+
+    fn validate(&self, definition: &PolicyDefinition) -> Result<(), Self::Error> {
+        let errors = definition
+            .statements()
+            .iter()
+            .flat_map(|statement| visit_statement(statement))
+            .collect::<Vec<_>>();
+
+        if !errors.is_empty() {
+            return Err(ValidatorError::ValidationSummary(errors));
+        }
         Ok(())
     }
 }
+
+fn visit_statement(statement: &Statement) -> Vec<String> {
```

Comment:
this Vec<String> looks suspicious as a list of errors

Reply:
it what way? any suggestions? not sure what you mean.

Reply:
I mean there is a method that returns `Vec<String>` and the last thing I'd imagine looking at the signature, that it will return a list of errors as strings.